api-spec/filters.md

```diff
@@ -0,0 +1,133 @@
+STAC Core is intending to provide a minimal set of features so that there is a low bar to entry for catalog
+implementations. The core concepts in STAC are Spatial and Temporal data, so those are the only filters we want
+to support in Core.
+
+Filters should be as simple as possible for a client to construct. Since we are returning json, we want to also
+consume json. This allows clients to support filtering without additional tools. So the core filters will be 
+JSON based. The search enpoint will accept application/json format queries, and GET on the collection will 
+support URL encoded JSON. POST search the is recommended way to filter results to avoid the URL encoding 
+issues that can happen with GET.
+
+Searching using POST will accept a JSON object where the top level keys are specifying which type of filter
+to apply. Those same top level key names can be used as query string parameters for the GET request. Items in the collection 
+should match all filters to be returned when querying. This implies a logical AND operation. If
+an OR operation is needed, it should be specified through an extension filter.
+
+For the spatial query, we are starting with the simplest form for a client to generate: a bbox. The bbox query 
+is a simple array of WGS84 Longitude and Latitude values. They are in this order: [west, south, east, north]. 
+This query will perform an intersects operation on the geometry values of the items in the catalog. Some GeoJSON 
+objects may provide a bbox property in addition to geometry, but it should not be used for the bbox filter since
+it is an optional field in GeoJSON.
+
+example
+```
+{
+  "bbox": [-180,-90,180,90]
+}
+```
+
+The temporal query will be based on ISO 8601 and should support time ranges as well as equality. To support range
+queries, we are using a simple JSON based language. Ranges will be specified as an object with keys indicating the comparison to use.
+
+Equality is specified as `{"time": "2018-03-20T16:11:44.353Z"}`  
+Before is `{"time":{"lt":"2018-03-20T16:11:44.353Z"}}`  
+After is `{"time":{"gt":"2018-03-20T16:11:44.353Z"}}`  
+Before with Equality is `{"time":{"lte":"2018-03-20T16:11:44.353Z"}}`  
+After with Equality is `{"time":{"gte":"2018-03-20T16:11:44.353Z"}}`  
+
+Filter Extensions
+-----------------
+
+Spatial and Temporal filtering is not enough for many use cases. STAC is designed to be extensible, so we encourage additional filters to be provided as extensions.
```

Comment:
I worry a bit about being so flexible that a core implementation is not that useful. Or that an extension has to be very explicitly defined. 

Like could we just say that simple fields (numbers + text) should offer comparisons? I'm just thinking of like the use case of searching for satellite imagery - date and time and cloud cover. Can we make it a bit easier on content profiles if we don't make them explicitly define the querying for cloud cover? Like I was sorta imagining that a content profile like 'earth observation' wouldn't have to provide a modified openapi document, unless they really wanted some crazy new filter. 

I'm more than happy for those who do mongo / elasticsearch to push back due to implementation pain. I can see arguments that it's better to explicitly define the search terms, instead of doing a blanket statement. But I at least want to hear that's the reason, and we're not just going for extensibility for the sake of it.

I'd like us to present an opinionated core, that actually makes decisions about what should be done. And I think I'm ok to start just saying they're 'required' WFS can be the super flexible thing, and we can point people at that if STAC API doesn't meet their purposes. And I believe we also can require a bit more from the STAC API, because we can fall back on static stac's for people who want to keep things really simple.

Reply:
I think querying on every property by default is a little too open ended. It doesn't put a clear set of boundaries to be "core" compliant with a strongly defined set of search parameters. And I also worry about the scalability issues of allowing users to mix and match any number of search criteria. Some implementations might support that easily, but do we make it a requirement that all implementations should support indexes on all properties? I know our first implementation is going to be mongo based and it requires indexes to be defined ahead of time.

I do want to be opinionated on the schema used for specific data types though. I think number fields should always use the NumberRange filter schema. I'm not yet sure how to document that other than making it a strongly worded recommendation.

I could see a "search all primitives" extension that says something like "use the propery name as the key, and the value should match the type: <table of type to filter schema>"